pkg/miner/miner.go

```diff
@@ -49,11 +50,21 @@ func (m *Miner) Mine(miningFn func()) {
 // Stop asks all worker functions to stop their work. Note that there is no
 // guarantee the worker function will stop immediately but the miner will not
 // call the worker function again until the miner's work is resumed.
+// Stop can be called multiple times. Each call increases a count on the
+// internal latch by one. To resume the work, Resume function needs to be as
+// many times as Stop(). This way, several goroutines can stop the computations
+// and they all need to agree on resuming them.
 func (m *Miner) Stop() {
+	m.latch.Add(1)
+
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
```

Comment:
Is this order of latch and mutex intentional (same in `Resume`)? In the current form, both operations (`Stop` and `Resume`) are not atomic. For example, if we trigger two goroutines, the first calling `Stop` and the second `Resume`, I think the following is possible:
1. G1 calls `Stop` and increments the latch's counter
2. G2 calls `Resume` and overtakes G1 by decrementing the counter and taking the mutex lock then makes some operations and unlocks the mutex.
3. G1 takes the mutex lock and performs its own operations

Although such a scenario is harmless in the current implementation, I think it's quite error-prone and may cause weird problems in the future. Shouldn't we always take the mutex first?

Reply:
With the change in this PR we want to resume the computations only when all goroutines that requested to stop them have _signalled_ that they completed their work and computations can be resumed.

The _signalled_ element is crucial. We really need all of them to tell the miner that the computations can be resumed. We do not know when the operations that are supposed to stop the computations are invoked and how long they are going to take. I was playing with an idea of a timeout but that turned out to be complicated and error-prone.

Since we need to trust goroutines to call `Resume`, it means we assume goroutines are working as expected. I am not totally in love with this assumption but I could not figure out anything else if we do not want the generator to just keep running in the background.

If we assume goroutines are working as expected and always call `Resume` after `Stop`, it also means we assume they are not going to call `Resume` twice because it breaks the assumption of goroutines being correctly implemented (this answers the question in another comment).

In the scenario you described, G2 would first need to call `Stop`:
1. G2 calls `Stop` and increments the latch's counter. Let's say it's preempted after the latch.
2. G1 calls `Stop` and increments the latch's counter. It's stopping computations
4. G1 calls `Resume` and blocks on the latch.
5. G2 is still in `Stop`. It acquires the lock and sees that the miner is already stopped so it returns.
6. G2 calls `Resume` and it's resuming computations.
7. G1 is still in `Resume`. It acquires the lock and sees that the miner is already working so it returns.

In the current setup, the lock is acquired only to actually stop or resume goroutines and after the lock is acquired we check the state. `Stop` and `Resume` are not atomic but I think everything should work fine under the assumption goroutines are:
1. Always calling `Stop()` before `Resume()`
2. They never forget to call `Resume()` if they called `Stop()`.

If we tried to make both `Stop()` and `Miner()` atomic and acquire the lock at the beginning, I think we would run into a deadlock:

1. G1 calls `Stop`, acquires the lock, increases the counter on the latch, and stops the computations.
2. G2 calls `Stop`, acquires the lock, and increased the counter on the latch.
3. G2 calls `Resume`, acquires the lock, decreases the counter on the latch, and blocks on wait.
4. G1 calls `Resume` but it can not acquire the lock, so it waits.

I am open to all suggestions on how to improve this mechanism. I could come up only with this solution if we want to stop the miner. I was also thinking about abandoning this path completely and reworking the pre-params generator code in tss-lib to always work in a single thread, even if generating pre-params will take ages but I don't think it's a better option. 4 years ago, when we were working on the safe prime generator and were optimizing the code pretty heavily, we [decided to go with concurrency](https://github.com/didiercrunch/paillier/blob/master/safe_prime_generator.go#L63) ([benchmarks](https://github.com/didiercrunch/paillier/pull/9#issuecomment-409124089)). What appeals to me about the current approach is that in a real world, we'll not have a thousands of goroutines on a single client. Just a few executing DKG or relay entry and the call to the miner is quite simple: `miner.Stop(); defer miner.Resume()`.

Reply:
(I have also updated the PR description)

Reply:
... And I can also clarify it in the code documentation in case we can't come up with anything better.

Reply:
Ok, that makes sense. Let's just make the assumptions about client goroutines super explicit in the documentation and mention that redundant calls to `Resume` will cause a panic.

Reply:
https://github.com/keep-network/keep-core/pull/3174/commits/e144844dcbf8dfaee89d23428a0c476907f76fd0